vividus-engine/src/main/java/org/vividus/variable/DynamicVariableCalculationResult.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2019-2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.vividus.variable;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+public final class DynamicVariableCalculationResult
+{
+    private final Optional<String> value;
+    private final Optional<String> error;
+
+    private DynamicVariableCalculationResult(Optional<String> value, Optional<String> error)
+    {
+        this.value = value;
+        this.error = error;
+    }
+
+    public static DynamicVariableCalculationResult withValue(String value)
+    {
+        return new DynamicVariableCalculationResult(Optional.of(value), Optional.empty());
+    }
+
+    public static DynamicVariableCalculationResult withError(String error)
+    {
+        return new DynamicVariableCalculationResult(Optional.empty(), Optional.of(error));
+    }
+
+    public static DynamicVariableCalculationResult withValueAndError(Optional<String> value,
```

Comment:
the "and" method seems a bit misleading for me because if the value is not empty the error is lost and never reachable in the code

Reply:
please propose a new name

Reply:
https://github.com/vividus-framework/vividus/pull/2376/files#r784737856

Reply:
or maybe like a named constructor `DynamicVariableCalculationResult.error(String error)`  `DynamicVariableCalculationResult.value(String value)`

Reply:
in generar I'm okay with describing this behaviour in javadoc